### PR TITLE
Views count

### PR DESCRIPTION
**Pull request by qasema**
fixed issue #856, add a view counter on the event view page to keep track of the people who view an event. 

## Testing
-  checkout branch viewsCount
-  reset the database 
-  make sure tests are not failing 
-  try to view an event (not admin)
-  switch the current user to admin and see if the view is counted 
-  only the admin can see the views, admin views do not count, and no duplicate views 
-  check if the EventView table is updated

### CODE REVIEW

**Review by solijonovam — COMMENTED** [review]
looks good to me!

**Review by solijonovam — COMMENTED** [review]
tests all pass! good job ala

File: app/logic/events.py [hunk 1]
```diff
@@ -450,3 +451,11 @@ def getTomorrowsEvents():
     tomorrowDate = date.today() + timedelta(days=1)
     events = list(Event.select().where(Event.startDate==tomorrowDate))
     return events
+def eventViewCount(viewer,event):
```
Comment (AndersonStettner) [inline app/logic/events.py]: We add a test for new logic functions.

**Review by AndersonStettner — CHANGES_REQUESTED** [review]

File: app/controllers/admin/routes.py [hunk 2]
```diff
@@ -147,7 +148,12 @@ def createEvent(templateid, programid=None):
 @admin_bp.route('/eventsList/<eventId>/view', methods=['GET'])
 @admin_bp.route('/eventsList/<eventId>/edit', methods=['GET','POST'])
 def eventDisplay(eventId):
-    # Validate given URL
```
Comment (AndersonStettner) [inline app/controllers/admin/routes.py:150]: Why did you remove this comment and not move it down with the content it was originally with?

File: app/controllers/admin/routes.py [hunk 3]
```diff
@@ -147,7 +148,12 @@ def createEvent(templateid, programid=None):
...
+    viewCount = EventView.select().where(EventView.event == eventId).count() # count the event views
```
Comment (AndersonStettner) [inline app/controllers/admin/routes.py]: Instead of having a trailing comment why not just change the name of the variable so it is clear by just reading its name.

File: app/controllers/admin/routes.py [hunk 4]
```diff
@@ -147,7 +148,12 @@ def createEvent(templateid, programid=None):
 @admin_bp.route('/eventsList/<eventId>/view', methods=['GET'])
 @admin_bp.route('/eventsList/<eventId>/edit', methods=['GET','POST'])
 def eventDisplay(eventId):
-    # Validate given URL
+    viewCount = EventView.select().where(EventView.event == eventId).count() # count the event views 
+    if not g.current_user.isCeltsAdmin and request.method == 'GET' and request.path == f'/eventsList/{eventId}/view':
+        viewer = User.get_by_id(g.current_user)
+        event = Event.get_by_id(eventId)
+        eventViewCount(viewer,event) #insert into the EventView table
```
Comment (AndersonStettner) [inline app/controllers/admin/routes.py]: There really is not a reason for this comment.

File: app/logic/events.py [hunk 5]
```diff
@@ -450,3 +451,11 @@ def getTomorrowsEvents():
...
+    """This checks if the current user already viewed the event. If not, insert a recored to EventView table"""
+    event_view_exists = EventView.select().where(EventView.user == viewer, EventView.event == event).exists()
```
Comment (AndersonStettner) [inline app/logic/events.py]: We use camleCase for variable naming/declaration across our logic files, please stay consistent.

File: app/logic/events.py [hunk 6]
```diff
@@ -450,3 +451,11 @@ def getTomorrowsEvents():
...
     events = list(Event.select().where(Event.startDate==tomorrowDate))
     return events
+def eventViewCount(viewer,event):
+    """This checks if the current user already viewed the event. If not, insert a recored to EventView table"""
+    event_view_exists = EventView.select().where(EventView.user == viewer, EventView.event == event).exists()     
+    if not event_view_exists:
+        # Insert new EventView record for the user and event
+        date = datetime.datetime.now()
+        new_event_view = EventView.create(user=viewer, event=event, viewedOn=date)
+        new_event_view.save()
```
Comment (AndersonStettner) [inline app/logic/events.py]: camelCase

File: app/templates/eventView.html [hunk 7]
```diff
@@ -123,5 +123,12 @@ <h3><br>Program Trainings</h3>
                     <button type="submit" class="btn btn-primary" value="{{eventData.id}}" id="rsvpBtn">RSVP</button>
                  {% endif %}
               {% endif %}
+              {% if g.current_user.isCeltsAdmin or isProgramManager %}
+            <div class="d-flex align-items-center">
+                <i class="bi bi-eye-fill me-2"></i>
+                <span class="me-2">Views:</span>
+                <span id="view-count">{{viewCount}}</span>
```
Comment (AndersonStettner) [inline app/templates/eventView.html]: We have been using camelCase for id names everywhere else, please keep it contestant.

File: app/templates/eventView.html [hunk 8]
```diff
@@ -123,5 +123,12 @@ <h3><br>Program Trainings</h3>
                     <button type="submit" class="btn btn-primary" value="{{eventData.id}}" id="rsvpBtn">RSVP</button>
                  {% endif %}
               {% endif %}
+              {% if g.current_user.isCeltsAdmin or isProgramManager %}
+            <div class="d-flex align-items-center">
+                <i class="bi bi-eye-fill me-2"></i>
+                <span class="me-2">Views:</span>
```
Comment (AndersonStettner) [inline app/templates/eventView.html]: Personally I would prefer it to be something like "Unique Page Views" or at least "Page Views".

**Review by AndersonStettner — CHANGES_REQUESTED** [review]

File: app/templates/eventView.html (hunk 8 above)
Comment (qasema) [reply app/templates/eventView.html]: agree

**Comment by qasema** [comment]
[Event Page views count](https://trello.com/c/mhqN5wfW/609-event-page-views-count)

**Comment by qasema** [comment]
[Add a view counter to event pages #856](https://trello.com/c/HdkV5aat/603-add-a-view-counter-to-event-pages-856)

**Review by sreynit02 — DISMISSED** [review]
Looks good!

File: app/controllers/admin/routes.py [hunk 9]
```diff
@@ -147,6 +148,11 @@ def createEvent(templateid, programid=None):
 @admin_bp.route('/eventsList/<eventId>/view', methods=['GET'])
 @admin_bp.route('/eventsList/<eventId>/edit', methods=['GET','POST'])
 def eventDisplay(eventId):
+    pageViewsCount = EventView.select().where(EventView.event == eventId).count() # count the event views
```
Comment (AndersonStettner) [inline app/controllers/admin/routes.py]: Removed comment, the variable name now gives accurate description of what it is doing so the comment is no longer needed.

File: app/controllers/admin/routes.py [hunk 10]
```diff
@@ -147,6 +148,11 @@ def createEvent(templateid, programid=None):
 @admin_bp.route('/eventsList/<eventId>/view', methods=['GET'])
 @admin_bp.route('/eventsList/<eventId>/edit', methods=['GET','POST'])
 def eventDisplay(eventId):
+    pageViewsCount = EventView.select().where(EventView.event == eventId).count() # count the event views 
+    if not g.current_user.isCeltsAdmin and request.method == 'GET' and request.path == f'/eventsList/{eventId}/view':
+        viewer = User.get_by_id(g.current_user)
```
Comment (AndersonStettner) [inline app/controllers/admin/routes.py]: `g.current_user` is the user so, you already have the user, making the `.get_by_id` unnecessary

File: app/logic/events.py (hunk 1 above)
Comment (AndersonStettner) [inline app/logic/events.py]: `eventViewCount` is a bit misleading of a name, something like `recordEventView`, `logEventView`, or `addEventView` would make more sense since that is what the function is doing.

File: app/logic/events.py [hunk 11]
```diff
@@ -450,3 +451,11 @@ def getTomorrowsEvents():
     tomorrowDate = date.today() + timedelta(days=1)
     events = list(Event.select().where(Event.startDate==tomorrowDate))
     return events
+def eventViewCount(viewer,event):
+    """This checks if the current user already viewed the event. If not, insert a recored to EventView table"""
+    eventViewExists = EventView.select().where(EventView.user == viewer, EventView.event == event).exists()     
+    if not eventViewExists:
+        # Insert new EventView record for the user and event
+        date = datetime.datetime.now()
```
Comment (AndersonStettner) [inline app/logic/events.py]: Use `.get_or_create()` instead of `.exists()` and ignore viewed on since you are adding default to model. Also, add the logic to exclude Admin into the function and add it to the tests.

File: tests/code/test_event_list.py [hunk 12]
```diff
@@ -192,3 +193,32 @@ def test_getOtherEvents(special_otherEvents):
...
+                                    bnumber = "B000000000",
+                                    email = "test@test.com",
+                                    phoneNumber = "Null",
+                                    firstName = "TestFirst",
+                                    lastName = "TestLast",
+                                    isStudent = False,
+                                    isFaculty = True,
+                                    isStaff = False,
+                                    isCeltsAdmin = False,
+                                    isCeltsStudentStaff = False)
```
Comment (AndersonStettner) [inline tests/code/test_event_list.py]: Indent `viewer `and `testEvent `properly.

File: tests/code/test_event_list.py [hunk 13]
```diff
@@ -192,3 +193,32 @@ def test_getOtherEvents(special_otherEvents):
...
+                            timeStart = "18:00:00",
+                            timeEnd = "21:00:00",
+                            location = "basement",
+                            isTraining = True,
+                            startDate = "2021-12-12",
+                            endDate = "2021-12-13")
+        viewer = User.create(username = "eventViewer",
+                                    bnumber = "B000000000",
+                                    email = "test@test.com",
+                                    phoneNumber = "Null",
```
Comment (AndersonStettner) [inline tests/code/test_event_list.py]: Give it a fake phone number. Just do something like 000-000-0000

File: app/models/eventViews.py [hunk 14]
```diff
@@ -0,0 +1,8 @@
+from app.models import*
+from app.models.user import User
+from app.models.event import Event
+
+class EventView(baseModel):
+    user = ForeignKeyField(User)
+    event = ForeignKeyField(Event)
+    viewedOn = DateTimeField()
```
Comment (AndersonStettner) [inline app/models/eventViews.py]: We should make `viewedOn` have a default of now.

File: app/templates/eventView.html [hunk 15]
```diff
@@ -123,5 +123,12 @@ <h3><br>Program Trainings</h3>
                     <button type="submit" class="btn btn-primary" value="{{eventData.id}}" id="rsvpBtn">RSVP</button>
                  {% endif %}
               {% endif %}
+              {% if g.current_user.isCeltsAdmin or isProgramManager %}
+            <div class="d-flex align-items-center">
+                <i class="bi bi-eye-fill me-2"></i>
+                <span class="me-2">Page Views:</span>
```
Comment (AndersonStettner) [inline app/templates/eventView.html]: Also, after thinking about it more, "Unique Viewers" would probably be the best descriptor.

**Review by AndersonStettner — CHANGES_REQUESTED** [review]

File: app/logic/events.py (hunk 11 above)
Comment (qasema) [reply app/logic/events.py]: I think that exclusion is happening in the controller. shall we add that extra layer in logic

File: app/templates/eventView.html (hunk 15 above)
Comment (qasema) [reply app/templates/eventView.html]: how about we make it like "Single User Views" because Unique Viewers could be a bit misleading

File: tests/code/test_event_list.py (hunk 13 above)
Comment (qasema) [reply tests/code/test_event_list.py]: should that necessarily make the test fail?

File: app/templates/eventView.html (hunk 15 above)
Comment (AndersonStettner) [reply app/templates/eventView.html]: How is it misleading? Unique Viewers seems to be a very accurate descriptor to me. It is a count of all unique users that have viewed the page.

File: tests/code/test_event_list.py (hunk 13 above)
Comment (AndersonStettner) [reply tests/code/test_event_list.py]: It is unclear right now whether you want to be empty or Null. Either way it is incorrect because this is not how you make a variable null in python.

File: app/logic/events.py (hunk 11 above)
Comment (AndersonStettner) [reply app/logic/events.py]: Move the check from the controller and put it into the logic and test it.

File: app/logic/events.py [hunk 16]
```diff
@@ -450,3 +451,6 @@ def getTomorrowsEvents():
     tomorrowDate = date.today() + timedelta(days=1)
     events = list(Event.select().where(Event.startDate==tomorrowDate))
     return events
+def addEventView(viewer,event):
+    """This checks if the current user already viewed the event. If not, insert a recored to EventView table"""
+    eventViewExists = EventView.get_or_create(user = viewer, event = event)
```
Comment (AndersonStettner) [inline app/logic/events.py]: The variable is irrelevant here.

**Review by AndersonStettner — CHANGES_REQUESTED** [review]

File: tests/code/test_event_list.py (hunk 13 above)
Comment (qasema) [reply tests/code/test_event_list.py]: but most of the previous test cases used are using the same thing. and I just used it.

File: app/templates/eventView.html (hunk 15 above)
Comment (qasema) [reply app/templates/eventView.html]: haha, I meant that Unique Viewers could refer to certain users and we refer to them as unique

File: tests/code/test_event_list.py (hunk 13 above)
Comment (AndersonStettner) [reply tests/code/test_event_list.py]: That is not an excuse, when things like this are noticed the person who is working on the pr/issue goes and fixes them.

File: app/templates/eventView.html (hunk 15 above)
Comment (AndersonStettner) [reply app/templates/eventView.html]: Brian and I agree that Unique Viewers is the best description here.

**Review by AndersonStettner — DISMISSED** [review]

**Review by BrianRamsay — APPROVED** [review]
